doc: Checkout libraries documentation for 5.0 (#1419)
SAP/spartacus-docs: _pages/home/release-notes/checkout-libraries-release-notes.md

```typescript
  constructor(
    ...
    protected checkoutDeliveryAddressFacade: CheckoutDeliveryAddressFacade,
    protected checkoutDeliveryModesFacade: CheckoutDeliveryModesFacade,
    ...
  ) {}
```

1. Facades execute commands and queries (example: [CheckoutDeliveryAddressFacade](https://github.com/SAP/spartacus/blob/develop/feature-libs/checkout/base/core/facade/checkout-delivery-address.service.ts))

Most of the time, each facade service injects both the `CommandService` and `QueryService`.
In `CheckoutQueryFacade` we have created a checkout details query which serves as a single source of truth for the checkout state, and is updated after each change the user does during the checkout.

A [command](https://sap.github.io/spartacus-docs/commands-and-queries/#commands-overview) is created in order to execute an action. In this case, it is for creating a delivery address. As seen in the snippet below, the command type is of `Command<Address, unknown>`. Address is for the input and unknown is for the output. The input for this case is required for the command to execute while the output 'may' be unknown.

1. Commands can dispatch events (for other side effects)

> The following snippet is taken from the `createDeliveryAddressCommand`, which can be found in the source code under the [checkout-delivery-address.service.ts](https://github.com/SAP/spartacus/blob/develop/feature-libs/checkout/base/core/facade/checkout-delivery-address.service.ts#L29). After successfully creating an address, we are [dispatching events](https://sap.github.io/spartacus-docs/event-service/#dispatching-individual-events) under the tap operator in order to be caught by our event listeners to perform additional actions.

Reviewer: ```suggestion
The following snippet is taken from the `createDeliveryAddressCommand`, which can be found in the source code under the [checkout-delivery-address.service.ts](https://github.com/SAP/spartacus/blob/7bba93e0e6b75024371361b169348cfaebeb2c7b/feature-libs/checkout/base/core/facade/checkout-delivery-address.service.ts#L29). After successfully creating an address, we are [dispatching events](https://sap.github.io/spartacus-docs/event-service/#dispatching-individual-events) under the tap operator in order to be caught by our event listeners to perform additional actions.
```

Use the "copy permalink" option, as it captures the file state at the moment of writing this doc.
It will likely evolve in the future, the lines will shift, and the doc won't be correct anymore.

<img width="1252" alt="Screenshot 2022-04-25 at 12 08 13" src="https://user-images.githubusercontent.com/11561723/165068946-89ce55f0-d3e0-4eb3-8461-aefb7e134fdc.png">

Author: thanks for sharing!

```typescript
return this.checkoutDeliveryAddressConnector
  .createAddress(userId, cartId, payload)
  .pipe(
    tap(() => {
      ...
    }),
    map((address) => {
      ...
    }),
    tap((address) =>

>       this.eventService.dispatch( <-- Event dispatching

Reviewer: ```suggestion
      this.eventService.dispatch( // <-- Event dispatching
```

```typescript
  protected onDeliveryAddressChange(): void {
    this.subscriptions.add(
      this.eventService
        .get(CheckoutDeliveryAddressCreatedEvent)
        .subscribe(({ userId, cartId }) => {
          this.globalMessageService.add(
            { key: 'addressForm.userAddressAddSuccess' },
            GlobalMessageType.MSG_TYPE_CONFIRMATION
          );

          this.eventService.dispatch(
            { userId, cartId },
            CheckoutResetDeliveryModesEvent
          );

          this.eventService.dispatch({}, CheckoutResetQueryEvent);
        })
    );
    ...
```

The example for `CheckoutResetQueryEvent` being fired is to reset our checkout details query from `CheckoutQueryFacade` due to the nature of our commands and queries, where we can [reload or reset a query](https://sap.github.io/spartacus-docs/commands-and-queries/#queries-overview).

### How to properly make use of queries in checkout

> Example: [CheckoutQueryFacade](https://github.com/SAP/spartacus/blob/develop/feature-libs/checkout/base/core/facade/checkout-query.service.ts) and [CheckoutDeliveryAddressComponent](https://github.com/SAP/spartacus/blob/develop/feature-libs/checkout/base/components/checkout-delivery-address/checkout-delivery-address.component.ts)

Reviewer: Better to use the permalink option when linking files

```typescript
  protected checkoutQuery$: Query<CheckoutState | undefined> =
    this.queryService.create<CheckoutState | undefined>(
      () =>
        this.checkoutPreconditions().pipe(
          switchMap(([userId, cartId]) =>
            this.checkoutConnector.getCheckoutDetails(userId, cartId)
          )
        ),
      {
        reloadOn: this.getQueryReloadTriggers(),
        resetOn: this.getQueryResetTriggers(),
      }
    );

    getCheckoutDetailsState(): Observable<QueryState<CheckoutState | undefined>> {
      return this.checkoutQuery$.getState();
    }
```

The getCheckoutDetailsState returns the entire state, which is an object composed of loading, error, and data properties, but you can always create additional methods in your extended service to include only getting the data. You would only need to use `.get()` instead of `.getState()` when executing queries.

```typescript
    getCheckoutDetails(): Observable<CheckoutState | undefined> {
      return this.checkoutQuery$.get();
    }
```

> When using queries are used to get the state, it is important to filter the loading state, and to consider using the `distinctUntilChanged` operator in order to reduce the number of emissions. Using the following snippet as an example, we are making sure to get the selected address when the data state is not loading. Moreover, we are making sure that we get distinct emissions instead to reduce the number of emissions.

Reviewer: ```suggestion
When queries are used to get the state, it is recommended to filter the loading state, and to consider using the RxJS' `distinctUntilChanged()` operator in order to reduce the number of emissions. Using the following snippet as an example, we are making sure to get the selected address when the data state is not loading. Moreover, we are making sure that we get distinct emissions instead to reduce the number of emissions.
```

    );
    ...
```

The example for [`CheckoutResetQueryEvent`](https://sap.github.io/spartacus-docs/event-service/#list-of-spartacus-events) being fired is to reset our checkout details query from `CheckoutQueryFacade` due to the nature of our commands and queries, where we can [reload or reset a query](https://sap.github.io/spartacus-docs/commands-and-queries/#queries-overview).

### How to properly make use of queries in checkout

Example: [CheckoutQueryFacade](https://github.com/SAP/spartacus/blob/7bba93e0e6b75024371361b169348cfaebeb2c7b/feature-libs/checkout/base/core/facade/checkout-query.service.ts) and [CheckoutDeliveryAddressComponent](https://github.com/SAP/spartacus/blob/7bba93e0e6b75024371361b169348cfaebeb2c7b/feature-libs/checkout/base/components/checkout-delivery-address/checkout-delivery-address.component.ts)

The 'checkout details query' that is injected in other checkout 'facade' services is used to get the current checkout details, which contains information on `delivery address, delivery mode, and payment information` when using base checkout. When using b2b checkout, you will have additional data like `purchase order number`, `cost centers`, and `payment type`.

> 

Reviewer 2: I would like to add a "lead-in" sentence here that says what this code snippet is, such as "The following is an example of the checkout details query".

Is it in fact an example of a checkout details query? If not, how can we refer to this example? What are we showing here? 🙂

Author: I think that lead-in makes sense

We're trying to show a checkout example on how-to use queries in checkout.

So we're saying: here is the 'facade' and a 'component' in our spartacus source repo. You should follow the 'flow' on how these are intertwined to make use of commands and query (specifically query)

---

> title: Checkout Libraries Release Notes

Reviewer: if using old components:
      1. you can consider using our new components. 
      2. If you don't want to, you can just import our new facade in your existing components, and keep using them and potentially slightly modifying them.

> title: Checkout Libraries Release Notes

Author: will add this later - keep this open

> title: Checkout Libraries Release Notes

Author: > cover the case when customers were using an old facade -> show how to switch to the new facade. This doesn't have to be a big section, as we'll have some migrations that'll handle this case.

isn't this covered by `### General idea of manually updating from old checkout facades in @spartacus/checkout/root`

> title: Checkout Libraries Release Notes

Reviewer: Yes, it is covered to a certain extent. I'll just expand it a bit

---

**Table of Contents**

- This will become a table of contents (this text will be scrapped).
  {:toc}

---

Originally, [Spartacus 4.0 released the new code splitted checkout library](https://sap.github.io/spartacus-docs/technical-changes-version-4/#new-checkout-library), which had only a single entrypoint that contained all checkout flows even if it would not be used, for example, [scheduling a replenishment](https://sap.github.io/spartacus-docs/scheduled-replenishment/#setting-up-a-replenishment-order). However, in Spartacus 5.0, we further decoupled the library into different business logic features in order to have a smaller bundle size.

In addition to creating different entrypoints to reduce the bundle size, we have mostly removed ngrx dependencies, with the exception of a few ngrx actions that are isolated within an event listener, by replacing them with [commands and queries](https://sap.github.io/spartacus-docs/commands-and-queries/#page-title). It is currently not possible to fully separate ngrx from checkout until other libraries from Spartacus get refactored, such as Cart or User libraries.

- Benefits of converting to `command and queries`:

>   - as all functionality is in classes, it is easier to extend, as opposed to ngrx where you can not really extend a reducer or an effect (without dismantling Spartacus' ngrx modules).

Author: makes sense 👍

>   - as all functionality is in classes, it is easier to extend, as opposed to ngrx where you can not really extend a reducer or an effect (without dismantling Spartacus' ngrx modules).

Author: makes sense 👍

  - commands are built in a more reactive way, and return the execution result as part of the same method call.

>   - similarly to the commands, listening to loading, error, and data state changes are as simple as calling one method and getting all the results in one call through queries.

Author: added, let me know what you think of it

>   - similarly to the commands, listening to loading, error, and data state changes are as simple as calling one method and getting all the results in one call through queries.

Reviewer: Looks good 👍🏻

  - page route has been also been changed from `/checkout/shipping-address` to `/checkout/delivery-address`
- CheckoutShippingAddressComponent `CMSComponent` has been renamed to CheckoutDeliveryAddress
- BodyContentSlot-checkoutShippingAddress `ContentSlot` has been renamed to BodyContentSlot-checkoutDeliveryAddress
- SideContent-checkoutShippingAddress `ContentSlot` has been renamed to SideContent-checkoutDeliveryAddress

Note: The new checkout library is not backwards compatible. This library is intended to be used with new applications that are created with other Spartacus 5.0 libraries. If you generate a new storefront app using schematics, the app will use the new checkout library by default. For more information, see [Integration Libraries and Feature Libraries]({{ site.baseurl }}/schematics/#integration-libraries-and-feature-libraries).

## Events

Checkout events are essential as they are used to cause additional side-effects. These events help remove the side effects done from `effects` when using ngrx. Events are caught either by event listeners or [queries to reset/reload data state](https://sap.github.io/spartacus-docs/commands-and-queries/#queries-overview),

> ### CheckoutReloadQueryEvent

Author: I added this in line 64, but were you trying to mention to not include the event explanation? I think we should include them because the list of spartacus events only shows path to file

> ### CheckoutReloadQueryEvent

Reviewer: Yep, we don't need to list them, as the list will be outdated at some point.
Customers are used to look at our source code anyways, and can see what each of the events are for.
Moreover, when we add / remove events to / from queries, we will document these changes in our migration doc, so the customers will be up to date

> ### CheckoutReloadQueryEvent

Author: removed list of events to not deal with evolving payload

```typescript
  constructor(
    ...
    protected checkoutDeliveryAddressFacade: CheckoutDeliveryAddressFacade,
    protected checkoutDeliveryModesFacade: CheckoutDeliveryModesFacade,
    ...
  ) {}
```

1. Facades execute commands and queries (example: [CheckoutDeliveryAddressFacade](https://github.com/SAP/spartacus/blob/develop/feature-libs/checkout/base/core/facade/checkout-delivery-address.service.ts))

Usually, each service injects both the `CommandService` and `QueryService`, but in the case of checkout, we have created a 'super query' called `CheckoutQueryFacade` that contains the usage of the `QueryService` as it is a common query for all checkout facade services in order to avoid repetition.

```typescript
  constructor(
    protected activeCartFacade: ActiveCartFacade,
    protected userIdService: UserIdService,
    protected eventService: EventService,
    protected commandService: CommandService,
    protected checkoutDeliveryAddressConnector: CheckoutDeliveryAddressConnector,
    protected checkoutQueryFacade: CheckoutQueryFacade
  ) {}

> ```

Author: removed! I agree, it will most likely change and this md would not reflect the current state of things

        switchMap(([userId, cartId]) => {
          return this.checkoutDeliveryAddressConnector
            .createAddress(userId, cartId, payload)
            .pipe(
              tap(() => {
                if (userId !== OCC_USER_ID_ANONYMOUS) {
                  this.eventService.dispatch(
                    { userId },
                    LoadUserAddressesEvent
                  );
                }
              }),
              map((address) => {
                address.titleCode = payload.titleCode;
                if (payload.region?.isocodeShort) {
                  address.region = {
                    ...address.region,
                    isocodeShort: payload.region.isocodeShort,
                  };
                }
                return address;
              }),
              tap((address) =>
                this.eventService.dispatch(
                  {
                    userId,
                    cartId,
                    address,
                  },
                  CheckoutDeliveryAddressCreatedEvent
                )
              )
            );
        })
      ),
    {
      strategy: CommandStrategy.CancelPrevious,
    }
  );

> ```

Author: got it

1. Commands can dispatch events (for other side effects)

> The following snippet is taken from the createDeliveryAddressCommand snipet above. After successfully creating an address, we have dispatched two [events](https://sap.github.io/spartacus-docs/event-service/#dispatching-individual-events) under the tap operator in order to be caught by our event listeners to perform additional actions.

Author: Hmmmmm If we're removing [above](https://github.com/SAP/spartacus-docs/pull/1419#discussion_r835263247), then this doesn't make sense anymore. 

I added a reference to our source code instead, specifically, the `checkout-delivery-address.service.ts`

```typescript
return this.checkoutDeliveryAddressConnector
  .createAddress(userId, cartId, payload)
  .pipe(
    tap(() => {
      if (userId !== OCC_USER_ID_ANONYMOUS) {
        this.eventService.dispatch( <-- 1st dispatched event
          { userId },
          LoadUserAddressesEvent
        );
      }
    }),
    map((address) => {
      address.titleCode = payload.titleCode;
      if (payload.region?.isocodeShort) {
        address.region = {
          ...address.region,
          isocodeShort: payload.region.isocodeShort,
        };
      }
      return address;
    }),
    tap((address) =>
      this.eventService.dispatch( <-- 2nd dispatched event
        {
          userId,
          cartId,
          address,
        },
        CheckoutDeliveryAddressCreatedEvent
      )
    )
```

> After an event that was dispatched during a command or query, our event listener that listened for [CheckoutDeliveryAddressCreatedEvent](#checkoutdeliveryaddresscreatedevent) fires a global message to let the user know that they have successfully created the delivery address.

Author: should we really remove this? It is a reference to the event explained under the events section, where it will automatically position the reader to the event

> After an event that was dispatched during a command or query, our event listener that listened for [CheckoutDeliveryAddressCreatedEvent](#checkoutdeliveryaddresscreatedevent) fires a global message to let the user know that they have successfully created the delivery address.

Reviewer: I think so, because we should remove the list of events from this doc.
You can maybe link a file using the permalink here, or link to the event docs (which is not yet updated, but maybe put a todo for Matt).

> After an event that was dispatched during a command or query, our event listener that listened for [CheckoutDeliveryAddressCreatedEvent](#checkoutdeliveryaddresscreatedevent) fires a global message to let the user know that they have successfully created the delivery address.

Author: I removed the TODO that references to put a reference for the event docs because we already have it in the generalized event section.

Moreover, the events there are automatically generated if I recall correctly

> Event listeners are flexible as we can inject any number of dependencies to perform certain actions. Moreover, in this case, it fires additional events, such as [CheckoutResetDeliveryModesEvent](#checkoutresetdeliverymodesevent) and [CheckoutResetQueryEvent](#checkoutresetqueryevent).

Author: hmmm. Leaving this open to ask if we should really remove the event references https://github.com/SAP/spartacus-docs/pull/1419#discussion_r835252663

          );

          this.eventService.dispatch(
            { userId, cartId },
            CheckoutResetDeliveryModesEvent
          );

          this.eventService.dispatch({}, CheckoutResetQueryEvent);
        })
    );
    ...
```

`CheckoutResetDeliveryModesEvent` is fired to re-load a cart, which ends up being caught by our `CheckoutLegacyStoreEventListener`. This event listener encapsulates the ngrx actions that we can not remove from checkout until other features changes to commands and queries.

The example for `CheckoutResetQueryEvent` being fired is to reset our super query from `CheckoutQueryFacade` due to the nature of our commands and queries, where we can [reload or reset a query](https://sap.github.io/spartacus-docs/commands-and-queries/#queries-overview).

### How to properly make use of queries in checkout

Example: [CheckoutQueryFacade](https://github.com/SAP/spartacus/blob/develop/feature-libs/checkout/base/core/facade/checkout-query.service.ts) and [CheckoutDeliveryAddressComponent](https://github.com/SAP/spartacus/blob/develop/feature-libs/checkout/base/components/checkout-delivery-address/checkout-delivery-address.component.ts)

> The 'super query' that is injected in other checkout 'facade' services is used to get the current checkout details, which contains information on `delivery address, delivery mode, and payment information` when using base checkout. When using b2b checkout, you will have additional data like `purchase order number`, `cost centers`, and `payment type`.

Author: I think checkout details query would suffice

```typescript
  protected checkoutQuery$: Query<CheckoutState | undefined> =
    this.queryService.create<CheckoutState | undefined>(
      () =>
        this.checkoutPreconditions().pipe(
          switchMap(([userId, cartId]) =>
            this.checkoutConnector.getCheckoutDetails(userId, cartId)
          )
        ),
      {
        reloadOn: this.getQueryReloadTriggers(),
        resetOn: this.getQueryResetTriggers(),
      }
    );

    getCheckoutDetailsState(): Observable<QueryState<CheckoutState | undefined>> {
      return this.checkoutQuery$.getState();
    }

> ```

Author: Removed, but I was thinking it would help the readers. But it really isn't necessary

> ```

Author: I think it's necessary as it helps with explaining the following things

The getCheckoutDetailsState returns the entire state, which is an object composed of loading, error, and data properties, but you can always create additional methods in your extended service to include only getting the data. You would only need to use `.get()` instead of `.getState()` when executing queries.

```typescript
    getCheckoutDetails(): Observable<CheckoutState | undefined> {
      return this.checkoutQuery$.get();
    }
```

## How to add additional logic to existing features

> TBD (wrappers / new schematics mechanism)

Reviewer: Hmm, maybe just point to the doc here. 
Instead of repeating ourselves for all features, we can just point to the general doc which explains how to do it

Author: which doc? 

I was going to add the docs after our schematics gets updated to support the wrapper modules

Reviewer: Ah, got it.
I was thinking about this doc: https://sap.github.io/spartacus-docs/lazy-loading-guide/#customizing-lazy-loaded-modules
I guess we will extract this example to a separate doc page.

Author: I found trails to this referred docs

https://github.com/SAP/spartacus/blob/develop/projects/schematics/README.md#glossary > https://github.com/SAP/spartacus/pull/15237 > https://wiki.one.int.sap/wiki/display/spar/Installing+extensions+for+an+existing+lazy+loaded+feature

btw, in our repo, it says to notify to update the link in this file, so letting you know now to update haha~

----
All in all, 

shouldn't this 'docs' be made available in the public spartacus `docs` repo as well, so that we can reference it here, which is why we have this TBD pending

Reviewer: Thanks for letting me know!
Yep, agreed. Until then, I linked to our existing doc which also explains how to extend LL features (if curious, please check the commits).
